chain-impl-mockchain/src/ledger/ledger.rs

```diff
@@ -723,66 +723,53 @@ impl Ledger {
         Ok(())
     }
 
-    /// Try to apply messages to a State, and return the new State if successful
-    pub fn apply_block(
+    /// This function must be called before applying fragments to the ledger. Do not use it if you
+    /// use `apply_block`.
+    pub fn apply_block_step1(
```

Comment:
It looks like this exposes an interim `Ledger` state in public API.

I think there should be a more ergonomic API: this method, renamed to, e.g. `begin_block()`, would return a wrapper type, say `BlockApply<'a>`.
That type would keep track of the metadata in a `&'a HeaderContentEvalContext` member, and likewise for the ledger parameters. An `fn apply_fragment(&mut self)` member should be present on this type, and the final method call, say, `fn finalize(self) -> Ledger`, would return the finallized block ledger state.